### PR TITLE
interchange: simplify protobuf decoding code

### PR DESCRIPTION
**Pull request by benesch**
Still need to shake out the testing here, so marking as a draft for now, but this is a fix for#8930 and also a major rewrite of our protobuf decoding pipeline. I tried to keep things separated into commits, but might ultimately be easier to review as if it's new code.

### CODE REVIEW

**Review by cjubb39 — COMMENTED** [review]
A couple inline qs that maybe are a little eager / better saved for later

Overall this seems much simpler, easier to follow, and closer to how we'd write it if we were to write it from scratch right now -- and it seems to have the advantage of 1) a more appropriate interface used in the `dataflow-types` crate; 2) correct nullability on columns in tests.  I'm also a fan of removing the need for some test-only things on `DecodedDescriptors`.

As far as testing goes, I agree with one of your commit messages that what we had here was mostly duplicative so I think just having the testdrive tests thoroughly check `derive_column_type` and `pack_{message, field, ...}` would be good.

Two shortcoming of those I believe current exist are the lack of
1) `repeated` non-Message types
2) exercising default values

**Comment by benesch** [comment]
Ok, @cjubb39, this is finally ready for review! This fixes up all the TODOs I left in the testdrive scripts, including the big one (#9250). Note that I re-rewrote the protobuf decoding logic to use the new `DynamicMessage` stuff in rust-protobuf directly, rather than going through serde-protobuf. This ends up being _even_ simpler (defaults are handled by rust-protobuf automatically, for example!), and is one less library to worry about.

I also ran the benchmark in `interchange` and observed a pretty sizable speedup. It's still not super fast, but it's not bad.

```
protobuf/decode         time:   [5.9428 us 5.9537 us 5.9659 us]                             
                        thrpt:  [30.532 MiB/s 30.595 MiB/s 30.651 MiB/s]
                 change:
                        time:   [-58.564% -58.480% -58.397%] (p = 0.00 < 0.05)
                        thrpt:  [+140.37% +140.85% +141.34%]
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

```

**Comment by benesch** [comment]
> As far as testing goes, I agree with one of your commit messages that what we had here was mostly duplicative so I think just having the testdrive tests thoroughly check `derive_column_type` and `pack_{message, field, ...}` would be good.
> 
> Two shortcoming of those I believe current exist are the lack of
> 
> 1. `repeated` non-Message types
> 2. exercising default values

Just for posterity's sake: I 100% agreed here, and sent #9240 separately to patch up those testing holes.

File: src/interchange/src/protobuf.rs [hunk 1]
```diff
@@ -103,436 +158,144 @@ fn derive_scalar_type<'a>(
...
+        FieldLabel::Required => bail!("Required field {} not supported", field.name()),
+        FieldLabel::Repeated => Ok(ColumnType {
+            nullable: false,
+            scalar_type: ScalarType::List {
+                element_type: Box::new(scalar_type),
+                custom_oid: None,
+            },
+        }),
+        FieldLabel::Optional => Ok(ColumnType {
+            nullable: matches!(field_type, FieldType::Message(_)),
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/interchange/src/protobuf.rs]: (this code's been rewritten, and I think the new structure addresses this nicely!)

**Comment by benesch** [comment]
Actually, going to get rid of the last commit that attempted to address #9250. The problem is much worse than I thought.

File: src/interchange/src/protobuf.rs [hunk 2]
```diff
@@ -103,436 +158,144 @@ fn derive_scalar_type<'a>(
...
-                    Datum::Float64(OrderedFloat(0.0)),
-                    Datum::String("RED"),
-                    Datum::String("one")
-                ]
-            )
-        } else {
-            panic!("Expected the first field to be a list of datums!");
+        Value::U32(_) | Value::U64(_) => {
+            bail!(
+                "internal error: unexpected value while decoding protobuf message: {:?}",
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/interchange/src/protobuf.rs]: Yeah exactly: #7629. A much nicer error gets generated during schema validation.

File: src/interchange/src/protobuf.rs [hunk 3]
```diff
@@ -103,436 +158,144 @@ fn derive_scalar_type<'a>(
...
+        Value::Bool(false) => packer.push(Datum::False),
+        Value::Bool(true) => packer.push(Datum::True),
+        Value::I32(i) => packer.push(Datum::Int32(i)),
+        Value::I64(i) => packer.push(Datum::Int64(i)),
+        Value::F32(f) => packer.push(Datum::Float32(f.into())),
+        Value::F64(f) => packer.push(Datum::Float64(f.into())),
+        Value::String(s) => packer.push(Datum::String(&s)),
+        Value::Bytes(b) => packer.push(Datum::Bytes(&b)),
+        Value::Enum(e) => {
+            let enum_desc = match field_desc.field_type(descriptors) {
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/interchange/src/protobuf.rs]: This validation happens inside of rust-protobuf so there's no need to do it here! The structure of rust-protobuf is nicer too, so there isn't awkwardly only one branch of the match that has to look in the field desc.

**Review by cjubb39 — APPROVED** [review]
lgtm! I'm especially happy with the new interface in `dataflow-types`

**Comment by benesch** [comment]
Woohoo! Thanks very much for suffering through this review.

**Pull request merged**